### PR TITLE
Consolidate search parser registries

### PR DESCRIPTION
**Pull request by rjernst**
Parsing a search request is currently split up among a number of
classes, using multiple public static methods, which take multiple
regstries of elements that may appear in the search request like query
parsers and aggregations. This change begins consolidating all this code
by collapsing the registries normally used for parsing search requests
into a single SearchRequestParsers class. It is also made available to
plugin services to enable templating of search requests.  Eventually all
of the actual parsing logic should move to the class, and the registries
should be hidden, but for now they are at least co-located to reduce the
number of objects that must be passed around.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/SearchModule.java [hunk 1]
```diff
@@ -477,14 +483,9 @@ public PipelineAggregationSpec addResultReader(String writeableName, Writeable.R
     @Override
     protected void configure() {
         if (false == transportClient) {
-            /*
-             * Nothing is bound for transport client *but* SearchModule is still responsible for settings up the things like the
-             * NamedWriteableRegistry.
-             */
             bind(IndicesQueriesRegistry.class).toInstance(queryParserRegistry);
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/SearchModule.java]: Now that you are binding SearchRequestParsers do we need to bing IndicesQueriesRegistry?

File: core/src/main/java/org/elasticsearch/search/SearchRequestParsers.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * A container for all parsers used to parse
+ * {@link org.elasticsearch.action.search.SearchRequest} objects from a rest request.
+ */
+public class SearchRequestParsers {
+    // TODO: this class should be renamed to SearchRequestParser, and all the parse
+    // methods split across RestSearchAction and SearchSourceBuilder should be moved here
+
+    // TODO: IndicesQueriesRegistry should be removed and just have the map of query parsers here
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/SearchRequestParsers.java]: Maybe add a note about making the fields private?

**Comment by nik9000** [comment]
Seems like a good start. LGTM.

**Comment by dadoonet** [comment]
Congrats for the 20000th issue/pr:)

File: core/src/main/java/org/elasticsearch/search/SearchModule.java (hunk 1 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/search/SearchModule.java]: Unfortunately not, for now. It is used in quite a few places...

**Pull request merged**